### PR TITLE
wip: separate RepoHasher and CommitHasher, basic integration of CodeL…

### PR DESCRIPTION
**Pull request by anatolykislov**
…ongevity

### CODE REVIEW

File: src/main/kotlin/app/hashers/CodeLongevity.kt [hunk 1]
```diff
@@ -83,30 +92,31 @@ class CodeLongevity(repoPath: String, tailRev: String) {
...
+        val emails = hashSetOf(localRepo.author.email)
+
+        val sum: MutableMap<String, Long> = emails.associate { Pair(it, 0L) }
+                                                  .toMutableMap()
+        val total: Int = codeLines.size
         for (line in codeLines) {
-            val author = line.from.commit.getAuthorIdent()
-            if (author.getEmailAddress() != email) {
+            val email = line.from.commit.authorIdent.emailAddress
+            if (!emails.contains(email)) {
```
Comment (sergey-surkov) [inline src/main/kotlin/app/hashers/CodeLongevity.kt]: We also need to compute average for entire repo. Remember the chart?

File: src/main/kotlin/app/hashers/CodeLongevity.kt [hunk 2]
```diff
@@ -186,8 +198,12 @@ class CodeLongevity(repoPath: String, tailRev: String) {
                 }
 
                 // If a file was deleted, then the new path is /dev/null.
-                val path = if (newPath != DiffEntry.DEV_NULL) newPath else oldPath
-                var lines = files.get(path)!!
+                val path = if (newPath != DiffEntry.DEV_NULL) {
```
Comment (sergey-surkov) [inline src/main/kotlin/app/hashers/CodeLongevity.kt]: why did this change? the original looked more compact.

File: src/main/kotlin/app/hashers/CodeLongevity.kt [hunk 3]
```diff
@@ -209,7 +226,9 @@ class CodeLongevity(repoPath: String, tailRev: String) {
                         for (idx in delStart .. delEnd - 1) {
                             tmpLines.add(RevCommitLine(commit, oldPath, idx))
                         }
-                        lines.addAll(delStart, tmpLines)
+                        // TODO(alex): Approve code change.
+                        // delStart index caused IndexOutOfBoundException.
+                        lines.addAll(tmpLines)
```
Comment (sergey-surkov) [inline src/main/kotlin/app/hashers/CodeLongevity.kt]: We really need unit tests, right @asurkov ?

File: src/main/kotlin/app/hashers/CommitHasher.kt [hunk 4]
```diff
@@ -213,32 +175,15 @@ class RepoHasher(private val localRepo: LocalRepo, private val api: Api,
...
-            // Delete missing commits. If found at least one common commit
-            // then next commits are not deleted because hash of a commit
-            // calculated including hashes of its parents.
-            val firstOverlapCommit = findFirstOverlappingCommit()
-            val deletedCommits = repo.commits
-                .takeWhile { it.rehash != firstOverlapCommit?.rehash }
-            deleteCommitsOnServer(deletedCommits)
-        }
-
+        // Delete missing local commits. If found at least one common commit
```
Comment (sergey-surkov) [inline src/main/kotlin/app/hashers/CommitHasher.kt]: This sounds confusing. You are deleting commits on server, not locally.

File: src/main/kotlin/app/hashers/RepoHasher.kt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    init {
+        if (!RepoHelper.isValidRepo(localRepo.path)) {
+            throw IllegalArgumentException("Invalid repo $localRepo")
+        }
+
+        println("Hashing $localRepo...")
+        val git = loadGit(localRepo.path)
+        try {
+            localRepo.parseGitConfig(git.repository.config)
```
Comment (sergey-surkov) [inline src/main/kotlin/app/hashers/RepoHasher.kt]: I really like how neat this function is.

File: src/main/kotlin/app/hashers/CodeLongevity.kt (hunk 2 above)
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CodeLongevity.kt]: Doesn't fit 80 limit.

File: src/main/kotlin/app/hashers/CommitHasher.kt (hunk 4 above)
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CommitHasher.kt]: Ok.

File: src/main/kotlin/app/hashers/CodeLongevity.kt (hunk 1 above)
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CodeLongevity.kt]: Will be in next pr very-very soon.

File: src/main/kotlin/app/hashers/CodeLongevity.kt (hunk 3 above)
Comment (asurkov) [reply src/main/kotlin/app/hashers/CodeLongevity.kt]: Absolutely. Btw, I always though that TODO(name) is about who added a todo item, not a person who's supposed to fix it.

Comment (anatolykislov) [reply src/main/kotlin/app/hashers/CodeLongevity.kt]: I think of this as a way to notify a person. Is there any conventions on that?

**Pull request merged**